lucene/core/src/java/org/apache/lucene/codecs/lucene80/IndexedDISI.java

```diff
@@ -0,0 +1,536 @@
+        for (int i = 0 ; i < rank.length ; i++) {
+          out.writeShort(rank[i]);
+        }
+        for (long word : buffer.getBits()) {
+          out.writeLong(word);
+        }
+      }
+    } else {
+      BitSetIterator it = new BitSetIterator(buffer, cardinality);
+      for (int doc = it.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = it.nextDoc()) {
+        out.writeShort((short) doc);
+      }
+    }
+  }
+
+  // Creates a DENSE rank-entry (the number of set bits up to a given point) for the buffer.
+  // One rank-entry for every 512 bits/8 longs for a total of 128 shorts.
+  private static short[] createRank(FixedBitSet buffer) {
+    final short[] rank = new short[128];
+    final long[] bits = buffer.getBits();
+    int bitCount = 0;
+    for (int word = 0 ; word < 1024 ; word++) {
+      if (word >> 3 << 3 == word) { // Every 8 longs
+        rank[word >> 3] = (short)bitCount;
+      }
+      bitCount += Long.bitCount(bits[word]);
+    }
+    return rank;
+  }
+
+  /**
+   * Writes the docIDs from it to out, in logical blocks, one for each 65536 docIDs in monotonically
+   * increasing gap-less order.
+   * @param it  the document IDs.
+   * @param out destination for the blocks.
+   * @throws IOException if there was an error writing to out.
+   */
+  static void writeBitSet(DocIdSetIterator it, IndexOutput out) throws IOException {
+    final long origo = out.getFilePointer(); // All jumps are relative to the origo
+    int totalCardinality = 0;
+    int blockCardinality = 0;
+    final FixedBitSet buffer = new FixedBitSet(1<<16);
+    long[] jumps = new long[ArrayUtil.oversize(100, Long.BYTES)];
+    jumps[0] = out.getFilePointer()-origo; // First block starts at index 0
+    int prevBlock = -1;
+    int jumpBlockIndex = 0;
+
+    for (int doc = it.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = it.nextDoc()) {
+      final int block = doc >>> 16;
+      if (prevBlock != -1 && block != prevBlock) {
+        // Track offset+index from previous block up to current
+        jumps = addJumps(jumps, out.getFilePointer()-origo, totalCardinality, jumpBlockIndex, prevBlock+1);
+        jumpBlockIndex = prevBlock+1;
+        // Flush block
+        flush(prevBlock, buffer, blockCardinality, out);
+        // Reset for next block
+        buffer.clear(0, buffer.length());
+        totalCardinality += blockCardinality;
+        blockCardinality = 0;
+      }
+      buffer.set(doc & 0xFFFF);
+      blockCardinality++;
+      prevBlock = block;
+    }
+    if (blockCardinality > 0) {
+      jumps = addJumps(jumps, out.getFilePointer()-origo, totalCardinality, jumpBlockIndex, prevBlock+1);
+      flush(prevBlock, buffer, blockCardinality, out);
+      buffer.clear(0, buffer.length());
+      prevBlock++;
+    }
+    final int lastBlock = prevBlock == -1 ? 0 : prevBlock;
+//    jumps = addJumps(jumps, out.getFilePointer(), totalCardinality, lastBlock, lastBlock+1);
+    // NO_MORE_DOCS is stored explicitly
```

Comment:
Currently (as you note further down this might be changed), the jump-tables are not always used. So we still need `NO_MORE_DOCS` for small-stepping through the bits?